File: distinct_subsequences.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        m = len(s)
        n = len(t)

        memo = {}
        def rec(i1, i2):
            if (i1,i2) in memo:
                return memo[(i1,i2)]
            if i2 == n:
                memo[(i1, i2)] = 1
                return 1
            if i1 == m:
                memo[(i1, i2)] = 0
                return 0

            res = 0
            if s[i1] == t[i2]:
                res += rec(i1 + 1, i2 + 1)
            res += rec(i1 + 1, i2)
            memo[(i1, i2)] = res
            return res
        return rec(0,0)
            

File: distinct_subsequences.cpp
class Solution {
public:
    vector<vector<int>> cache;
    int dp(int i, int j, string& s, string & t) {
        if (i == s.size()) return j == t.size();
        if (j == t.size()) return 1;
        if (cache[i][j] != -1) return cache[i][j];
        int res = 0;
        if (s[i] == t[j]) res += dp(i+1, j+1, s, t);
        res += dp(i+1, j, s, t);
        return cache[i][j]=res;
    }

    int numDistinct(string s, string t) {
        cache.resize(s.size(), vector<int>(t.size(), -1));
        if (s.size() < t.size()) return 0;
        return dp(0,0,s,t);
    }
};
